### PR TITLE
Remove old timeout when deadline changes

### PR DESCRIPTION
**Pull request by yawkat**
I had a memory issue and @franz1981 suggested #211 as the cause. This patch is my fix for that bug, though I don't believe my mem issue was ultimately caused by this.

This PR does the legwork for adding ioringOpTimeoutRemove, and implementing a test. However two things can still be improved:

- [ ] could use IORING_TIMEOUT_UPDATE (see #211) to save one sqe.
- [x] there may be a race in IOUringEventLoop between the addTimeout and the IORING_OP_TIMEOUT handler. If the kernel fires a deadline cqe, then we send a deadline update sqe, and only then we process the first cqe, prevDeadlineNanos is NONE even though we've submitted a new deadline. I'm not sure if this can actually happen since deadline changes should only adjust the deadline downwards, not upwards? Not sure.

### CODE REVIEW

File: transport-native-io_uring/src/test/java/io/netty/incubator/channel/uring/NativeTest.java [hunk 1]
```diff
@@ -142,6 +143,64 @@ public void handle(int fd, int res, int flags, byte op, short mask) {
...
+            public void run() {
+                completionQueue.ioUringWaitCqe();
+                completionQueue.process(new IOUringCompletionQueueCallback() {
+                    int i = 0;
+
+                    @Override
+                    public void handle(int fd, int res, int flags, byte op, short mask) {
+                        if (i == 0) {
+                            assertEquals(Native.IORING_OP_TIMEOUT, op);
+                            assertEquals(-125, res);
```
Comment (normanmaurer) [inline transport-native-io_uring/src/test/java/io/netty/incubator/channel/uring/NativeTest.java:168]: what is `-125` here ?

Comment (yawkat) [reply transport-native-io_uring/src/test/java/io/netty/incubator/channel/uring/NativeTest.java:168]: -ECANCELED

Comment (normanmaurer) [reply transport-native-io_uring/src/test/java/io/netty/incubator/channel/uring/NativeTest.java:168]: lets add a comment

Comment (yawkat) [reply transport-native-io_uring/src/test/java/io/netty/incubator/channel/uring/NativeTest.java:168]:
```suggestion
                            // -ECANCELED
                            assertEquals(-125, res);
```

**Comment by normanmaurer** [comment]
> I had a memory issue and @franz1981 suggested #211 as the cause. This patch is my fix for that bug, though I don't believe my mem issue was ultimately caused by this.
> 
> This PR does the legwork for adding ioringOpTimeoutRemove, and implementing a test. However two things can still be improved:
> 
> * [ ]  could use IORING_TIMEOUT_UPDATE (see [Not Fired timers leak into the kernel #211](https://github.com/netty/netty-incubator-transport-io_uring/issues/211)) to save one sqe.
> * [ ]  there may be a race in IOUringEventLoop between the addTimeout and the IORING_OP_TIMEOUT handler. If the kernel fires a deadline cqe, then we send a deadline update sqe, and only then we process the first cqe, prevDeadlineNanos is NONE even though we've submitted a new deadline. I'm not sure if this can actually happen since deadline changes should only adjust the deadline downwards, not upwards? Not sure.

Thanks a lot ! I think what you did is fine... I would defer the use of IORING_TIMEOUT_UPDATE as it would also most likely bump the kernel version requirement. Maybe we can add an issue to keep track of it

**Comment by normanmaurer** [comment]
@yawkat so with this said I think we can put this out of draft state ?

**Comment by normanmaurer** [comment]
Also seems like the new added test failed :D

**Comment by yawkat** [comment]
i will look at the test but it might take a few days.

i think the race condition is still an issue. id like to at least think about it a bit more, consider how it can happen.

**Comment by yawkat** [comment]
ive introduced a "generation counter". basically, every addTimeout counts up the generation and sends it to the kernel as the extra field. When a timeout occurs, i check that the generation matches the latest timeout that was added. This ensures that a previous timeout cannot make us think that the current timeout is finished.

**Comment by franz1981** [comment]
well done @yawkat ! I like the epoch/sequencing approach

File: transport-classes-io_uring/src/main/java/io/netty/incubator/channel/uring/IOUringEventLoop.java [hunk 2]
```diff
@@ -248,9 +261,11 @@ private void handle(int fd, int res, int flags, byte op, short data) {
             pendingWakeup = false;
             addEventFdRead(ringBuffer.ioUringSubmissionQueue());
         } else if (op == Native.IORING_OP_TIMEOUT) {
-            if (res == Native.ERRNO_ETIME_NEGATIVE) {
+            if (res == Native.ERRNO_ETIME_NEGATIVE && data == prevTimeoutGeneration) {
                 prevDeadlineNanos = NONE;
```
Comment (franz1981) [inline transport-classes-io_uring/src/main/java/io/netty/incubator/channel/uring/IOUringEventLoop.java:265]: we could trace loging it too (as we do we fd reuse :( ) if we assume to be an unwelcome behaviour

**Comment by normanmaurer** [comment]
@yawkat seems like there is still a test failure

**Comment by yawkat** [comment]
thats weird. it passes locally every time, doesnt appear flakey.

**Comment by normanmaurer** [comment]
> thats weird. it passes locally every time, doesnt appear flakey.

did you run it in a loop ?

**Comment by normanmaurer** [comment]
@yawkat ping

**Comment by yawkat** [comment]
fixed

**Comment by normanmaurer** [comment]
@yawkat so I think this is now ready to be merged... correct ?

**Comment by yawkat** [comment]
yep, only potential improvement would be to use UPDATE, but not in this pr

**Comment by normanmaurer** [comment]
@yawkat ok perfect... Can you just update the commit message to match our template ?

**Comment by yawkat** [comment]
Can you squash merge it with this message:

---

Remove old timeout when deadline changes (#211)
Motivation:

When the event loop timer deadline changes, the old timer would not be removed. This could potentially lead to memory issues.

Modification:

Use IORING_OP_TIMEOUT_REMOVE to remove a previously set timeout before a new one is registered. Use a generation counter to prevent race conditions between timeout add/remove in sqe/cqe.

Result:

Old timers are removed. At any time, there is only at most one timer per event loop.

**Comment by normanmaurer** [comment]
@yawkat done... thanks a lot !

**Comment by yawkat** [comment]
thanks!

**Pull request merged**